### PR TITLE
Add more information to the contributing section especially laying out concerns for how to develop in singularity-eos and expectations for code review

### PR DESCRIPTION
**Pull request by Yurlungur**
## PR Summary

I believe a recent code review caused some friction. As such I want to lay out in writing some best practices and expectations for contributing to `singularity-eos`. This includes:
1. Some development concerns that should be on one's mind as one contributes to the project.
2. Expectations for a contributor undergoing code review.
3. Expectations for the person reviewing a pull request.

To be clear no one is at fault. However, I think having some expectations in writing that we can all point to and follow will help us interact more productively with each other going forward.

I'd like to reach out to the core team to ask for feedback on these expectations and then let's all try to follow them.

## PR Checklist

- [x] Adds a test for any bugs fixed. Adds tests for new features.
- [x] Format your changes by using the `make format` command after configuring with `cmake`.
- [x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,33 +2,24 @@
 
 ## Current develop
 
-### Updated `hdf5` handling in the build
-
-### Fixed (Repair bugs, etc)
-
```
Comment (Yurlungur) [inline CHANGELOG.md:8]: I think git munged the formatting of the change log at some point. I tried to clean it up here since I noticed it.

**Review by ktsai7 — APPROVED** [review]

File: doc/sphinx/src/contributing.rst [hunk 2]
```diff
@@ -37,6 +37,81 @@ internal tests will not be run automatically. So when the code is
...
+
+Expectations of code reviewers
+````````````````````````````````
+
+A good code review builds a contribution up, rather than tearing it
+down. Here are a few rules to keep code reviews constructive and
+congenial:
+
+* You should take the time needed to review a contribution and offer
+  meaningful advice. Unless a contribution is very small limit
```
Comment (mauneyc-LANL) [inline doc/sphinx/src/contributing.rst:86]: a comma after "very small"

File: doc/sphinx/src/contributing.rst [hunk 3]
```diff
@@ -50,6 +125,119 @@ and we'll engage with you to figure out how to proceed.
...
+    #. Heap memory allocation should only be performed at
+       initialization. Heap memory de-allocation should only be
+       performed at cleanup.
+
+    #. Restrict the length of functions to a single printed page.
+
+    #. Restrict the scope of data to the smallest possible.
+
+    #. Use the preprocessor sparingly. (The same applies for
+       non-trivial template metaprogramming.)
```
Comment (mauneyc-LANL) [inline doc/sphinx/src/contributing.rst:165]: suggestion to strike the parenthetical

1. it is imprecise; templates are a language feature, & they are resolved at compilation; they are very different from preprocessing.
2. the cat is out-of-the-bag here; `visit`/`variant` & so on

I don't insist on the change

Comment (Yurlungur) [reply doc/sphinx/src/contributing.rst:165]: No that's fair. I'll change it.

File: doc/sphinx/src/contributing.rst [hunk 4]
```diff
@@ -50,6 +125,118 @@ and we'll engage with you to figure out how to proceed.
...
+       they come up.
+
+.. _ten rules for developing safety-critical code: http://web.eecs.umich.edu/~imarkov/10rules.pdf
+
+* ``singularity-eos`` is a modern C++ code
+  and both standard template library capabilities and template
+  metaprogramming are leveraged frequently. This can sometimes make
+  parsing the code difficult. If you see something you don't
+  understand, ask. It may be it can be refactored to be more simple or
+  better documented.
```
Comment (jhp-lanl) [inline doc/sphinx/src/contributing.rst:179]: Looks like some typos here

File: doc/sphinx/src/contributing.rst [hunk 5]
```diff
@@ -50,6 +125,118 @@ and we'll engage with you to figure out how to proceed.
...
+
+* **``ports-of-call`` and portability decorators:** Functions that
+  should be run on device needs to be decorated with one of the
+  following macros: ``PORTABLE_FUNCTION``,
+  ``PORTABLE_INLINE_FUNCTION``,
+  ``PORTABLE_FORCEINLINE_FUNCTION``. These macros are imported from
+  the `ports-of-call`_ library and resolve to the appropriate
+  decorations for a given device-side backend such as cuda so the code
+  compiles correctly. Code that doesn't need to run on device does not
+  need these decorations.
```
Comment (jhp-lanl) [inline doc/sphinx/src/contributing.rst:198]: Maybe add an example that constructors aren't called on device?

**Review by jhp-lanl — APPROVED** [review]
Thank you @Yurlungur for this addition! I think these are some really good points and I appreciate them being written down.

File: doc/sphinx/src/contributing.rst (hunk 4 above)
Comment (Yurlungur) [reply doc/sphinx/src/contributing.rst:179]: I must have blindness to my own writing, since I don't see the typos. Can you point them out?

Comment (jhp-lanl) [reply doc/sphinx/src/contributing.rst:179]: Never mind! I think I read it out of order when I saw it last lol. My mistake!

**Review by junghans — APPROVED** [review]

**Comment by Yurlungur** [comment]
I think this is enough approvals. I'm going to merge it now. Thanks everyone for looking through this!

**Comment by Yurlungur** [comment]
If anyone looks at this after the fact and would like a change, please let me know. This should be a "living" document.

**Pull request merged**